ballista/rust/scheduler/src/scheduler_server/mod.rs

```diff
@@ -0,0 +1,142 @@
+
+// include the generated protobuf source as a submodule
+#[allow(clippy::all)]
+pub mod externalscaler {
+    include!(concat!(env!("OUT_DIR"), "/externalscaler.rs"));
+}
+
+mod event_loop;
+mod external_scaler;
+mod grpc;
+mod task_scheduler;
+
+type ExecutorsClient = Arc<RwLock<HashMap<String, ExecutorGrpcClient<Channel>>>>;
+
+#[derive(Clone)]
+pub struct SchedulerServer<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> {
+    pub(crate) state: Arc<SchedulerState<T, U>>,
+    pub start_time: u128,
+    policy: TaskSchedulingPolicy,
+    executors_client: Option<ExecutorsClient>,
+    event_loop: Option<Arc<EventLoop<SchedulerServerEvent>>>,
+    ctx: Arc<RwLock<ExecutionContext>>,
+    codec: BallistaCodec<T, U>,
+}
+
+impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> SchedulerServer<T, U> {
+    pub fn new(
+        config: Arc<dyn ConfigBackendClient>,
+        namespace: String,
+        ctx: Arc<RwLock<ExecutionContext>>,
+        codec: BallistaCodec<T, U>,
+    ) -> Self {
+        SchedulerServer::new_with_policy(
+            config,
+            namespace,
+            TaskSchedulingPolicy::PullStaged,
+            ctx,
+            codec,
+        )
+    }
+
+    pub fn new_with_policy(
+        config: Arc<dyn ConfigBackendClient>,
+        namespace: String,
+        policy: TaskSchedulingPolicy,
+        ctx: Arc<RwLock<ExecutionContext>>,
+        codec: BallistaCodec<T, U>,
+    ) -> Self {
+        let state = Arc::new(SchedulerState::new(config, namespace, codec.clone()));
+
+        let (executors_client, event_loop) =
+            if matches!(policy, TaskSchedulingPolicy::PushStaged) {
+                let executors_client = Arc::new(RwLock::new(HashMap::new()));
+                let event_action: Arc<SchedulerServerEventAction<T, U>> =
+                    Arc::new(SchedulerServerEventAction::new(
+                        state.clone(),
+                        executors_client.clone(),
+                    ));
+                let event_loop =
+                    Arc::new(EventLoop::new("scheduler".to_owned(), 10000, event_action));
+                (Some(executors_client), Some(event_loop))
+            } else {
+                (None, None)
+            };
+        Self {
+            state,
+            start_time: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_millis(),
+            policy,
+            executors_client,
+            event_loop,
+            ctx,
+            codec,
+        }
+    }
+
+    pub async fn init(&mut self) -> Result<()> {
+        {
+            // initialize state
+            let ctx = self.ctx.read().await;
+            self.state.init(&ctx).await?;
+        }
+
+        {
+            if let Some(event_loop) = self.event_loop.as_mut() {
+                // It's OK here, since we are sure the mutable reference only be used in the initialization
+                unsafe {
```

Comment:
Actually, in this PR the EventLoop don't need Arc. However, from the whole processing figure in #1704, the event loop will be used in somewhere else. The related PR is under progress.

Reply:
I see. In that case, does the `EventLoop` need to be passed around, or can we just pass a `Sender` to different producers? In that case, the `EventLoop` could still be owned and multiple producers could hold a clone of the event loops `Sender`. 

Maybe @alamb could comment on how much we should care to avoiding `unsafe`.

Reply:
The philosophy in Rust in general is to avoid `unsafe` as much as possible. Typically it `unsafe` is used for one of two reasons:
1. Interfacing with other languages (like `C/C++`) that don't have the same lifetime system as Rust
2. Performance where the overhead of doing things like array bounds checks is too great

Neither of those situations seems to apply here. 

I would personally suggest figuring out some way to avoid this `unsafe` block - @thinkharderdev  has one suggestion; 

I don't know this code well enough to offer anything specific, but I would say another pattern for something like this is "Interior Mutability" which is a fancy rust way of saying use something like `Mutex`. Perhaps

```rust
event_loop: Option<Arc<Mutex<EventLoop<SchedulerServerEvent>>>>,
```

Reply:
Hi @alamb, the reason I prefer not to add Mutex here is that the mutation only happens in the initialization phase. In other places, only it's only for reading. And we can avoid the overhead of locks.

@thinkharderdev, this PR is for providing some encapsulation for the channel processing and extracting some common behaviors in the EventLoop.

Reply:
Right, I think it's a great idea to encapsulate the event loop, but I think we can avoid the `unsafe` and the overhead of a `Mutex` or `RwLock` by decoupling the producer and consumer interface. If we want to encapsulate the error handling around sending an event on the underlying channel then perhaps we can create an explicit `EventProducer` like:

```rust
pub struct EventProducer<E> {
  tx_event: mpsc::Sender<E>
}

impl<E> EventProducer<E> {
      pub async fn post_event(&self, event: E) -> Result<()> {
        if let Some(tx_event) = self.tx_event.as_ref() {
            tx_event.send(event).await.map_err(|e| {
                BallistaError::General(format!("Fail to send event due to {}", e))
            })?
        } else {
            warn!("Event sender not exist!!!");
        }

        Ok(())
    }
} 

impl<E> EventLoop<E> {
  pub fn producer(&self) -> Result<EventProducer<E>> {
    EventPrdocuer::new(self.tx_event.cloned().ok_or(...)?)
  }
}
```

Reply:
My personal recommendation is go with the `safe` approach (even if it seems slower at first as it may require locks). My rationale is:

I have spent countless hours tracking down strange, subtle and hard to reproduce bugs related to memory corruption when working on C/C++ systems. My Rust experience has largely been free of such pain and I think the avoidance of that pain a key advantage of Rust and makes some of the painful parts of rust (like the borrow checker) worthwhile.

I also think the overhead of taking a mutex is likely to be low, and you could make it lower still by using using a `RwLock` instead. 

Even better would be if you can refactor the code so that it is clear when the initialization has occured so you need neither mutex nor unsafe, perhaps as @thinkharderdev  is suggesting

I like to think in most matters I am pretty pragmatic, but the use of `unsafe` is something I do feel quite strongly about.

Reply:
Thanks @alamb and @thinkharderdev. I think to avoid unsafe here, the only way is to use RwLock or Mutex. Although I still think it's not necessary, I will change it to use RwLock to avoid unsafe.

Reply:
Thanks @thinkharderdev and @alamb. I'll refine the code according to the @thinkharderdev's suggestion.